File: functions/src/engine/balootCard.ts
export const BALOOT_SUITS: BalootSuit[] = ['spades', 'hearts', 'diamonds', 'clubs'];
export const BALOOT_RANKS: BalootRank[] = ['7', '8', '9', '10', 'J', 'Q', 'K', 'A'];

export type BalootSuit = 'spades' | 'hearts' | 'diamonds' | 'clubs';
export type BalootRank = '7' | '8' | '9' | '10' | 'J' | 'Q' | 'K' | 'A';

export const SUIT_SYMBOLS: Record<BalootSuit, string> = {
  spades: '♠',
  hearts: '♥',
  diamonds: '♦',
  clubs: '♣',
};

const SUIT_CODES: Record<BalootSuit, string> = {
  spades: 'S',
  hearts: 'H',
  diamonds: 'D',
  clubs: 'C',
};

export class BalootCard {
  readonly suit: BalootSuit;
  readonly rank: BalootRank;

  constructor(suit: BalootSuit, rank: BalootRank) {
    this.suit = suit;
    this.rank = rank;
  }

  static fromString(key: string): BalootCard {
    const trimmed = key.trim();
    // Try the symbolic form first: A♠, 10♦
    for (const suit of BALOOT_SUITS) {
      const sym = SUIT_SYMBOLS[suit];
      if (trimmed.endsWith(sym)) {
        const rankPart = trimmed.slice(0, -sym.length);
        const rank = parseRank(rankPart);
        return new BalootCard(suit, rank);
      }
    }
    // Fallback to code form: AS, 10D
    if (trimmed.length >= 2) {
      const rankPart = trimmed.slice(0, -1);
      const code = trimmed.slice(-1).toUpperCase();
      const suit = Object.entries(SUIT_CODES).find(([, v]) => v === code)?.[0] as BalootSuit | undefined;
      if (suit) {
        return new BalootCard(suit, parseRank(rankPart));
      }
    }
    throw new Error(`Invalid BalootCard key: ${key}`);
  }

  get key(): string {
    return `${this.rank}${SUIT_SYMBOLS[this.suit]}`;
  }

  get code(): string {
    return `${this.rank}${SUIT_CODES[this.suit]}`;
  }

  equals(other: BalootCard): boolean {
    return this.suit === other.suit && this.rank === other.rank;
  }

  toString(): string {
    return this.key;
  }
}

function parseRank(part: string): BalootRank {
  const p = part.trim();
  if (p === '10') return '10';
  if (p === 'J' || p === 'j') return 'J';
  if (p === 'Q' || p === 'q') return 'Q';
  if (p === 'K' || p === 'k') return 'K';
  if (p === 'A' || p === 'a') return 'A';
  if (p === '7') return '7';
  if (p === '8') return '8';
  if (p === '9') return '9';
  throw new Error(`Invalid Baloot rank: ${part}`);
}

File: lib/features/game/domain/engine/baloot_card.dart
import 'package:flutter/foundation.dart';

import 'package:bloot/features/game/presentation/widgets/playing_card/playing_card_model.dart';

/// The four suits used in 32-card Saudi Baloot.
enum BalootSuit {
  spades(symbol: '♠', code: 'S'),
  hearts(symbol: '♥', code: 'H'),
  diamonds(symbol: '♦', code: 'D'),
  clubs(symbol: '♣', code: 'C');

  const BalootSuit({required this.symbol, required this.code});

  final String symbol;
  final String code;

  bool get isRed => this == hearts || this == diamonds;

  static BalootSuit fromCode(String code) =>
      values.firstWhere((s) => s.code == code.toUpperCase(), orElse: () => fromSymbol(code));

  static BalootSuit fromSymbol(String symbol) =>
      values.firstWhere((s) => s.symbol == symbol, orElse: () => throw FormatException('Invalid suit: $symbol'));

  CardSuit toCardSuit() => CardSuit.fromSymbol(symbol);
}

/// The eight ranks used in 32-card Saudi Baloot.
enum BalootRank {
  seven(label: '7'),
  eight(label: '8'),
  nine(label: '9'),
  ten(label: '10'),
  jack(label: 'J'),
  queen(label: 'Q'),
  king(label: 'K'),
  ace(label: 'A');

  const BalootRank({required this.label});

  final String label;

  static BalootRank fromLabel(String label) =>
      values.firstWhere((r) => r.label == label, orElse: () => throw FormatException('Invalid rank: $label'));

  CardRank toCardRank() => CardRank.fromLabel(label);
}

/// Immutable value object representing a single 32-card Baloot card.
@immutable
class BalootCard {
  const BalootCard({required this.suit, required this.rank});

  factory BalootCard.fromString(String card) {
    if (card.length < 2) throw FormatException('Invalid card string: $card');
    final suitSymbol = card.substring(card.length - 1);
    final rankLabel = card.substring(0, card.length - 1);
    return BalootCard(
      suit: BalootSuit.fromSymbol(suitSymbol),
      rank: BalootRank.fromLabel(rankLabel),
    );
  }

  factory BalootCard.fromJson(Object json) {
    final s = json as String;
    return BalootCard.fromString(s);
  }

  final BalootSuit suit;
  final BalootRank rank;

  String get key => '${rank.label}${suit.symbol}';

  bool get isRed => suit.isRed;

  PlayingCard toPlayingCard() => PlayingCard(rank: rank.toCardRank(), suit: suit.toCardSuit());

  static BalootCard fromPlayingCard(PlayingCard card) => BalootCard(
        rank: BalootRank.fromLabel(card.rank.label),
        suit: BalootSuit.fromSymbol(card.suit.symbol),
      );

  Object toJson() => key;

  @override
  String toString() => key;

  @override
  bool operator ==(Object other) =>
      identical(this, other) || other is BalootCard && suit == other.suit && rank == other.rank;

  @override
  int get hashCode => suit.hashCode ^ rank.hashCode;
}
